libcomposefs/lcfs-writer.c

```diff
@@ -1677,7 +1677,7 @@ int lcfs_node_unset_xattr(struct lcfs_node_s *node, const char *name)
 		assert(node->xattr_size >= 0);
 	}
 
-	return -1;
+	return 0;
```

Comment:
I think this change is wrong.  This -1 is the case where we didn't find the xattr to remove.  Arguably we should set `errno = ENODATA` here (which is the same as the `removexattr()` syscall.

The thing that's missing is the `return 0` on the "good path" inside the `if() { }` block, just above.

Reply:
> I think this change is wrong. 

I think it's the original code that's wrong.

>  This -1 is the case where we didn't find the xattr to remove. Arguably we should set errno = ENODATA here (which is the same as the removexattr() syscall.
> The thing that's missing is the return 0 on the "good path" inside the if() { } block, just above.

The API you propose makes sense to me and it does seem theoretically valuable for a caller to be able to distinguish "value not found" from other errors. But I'd note that it complicates the desired logic for the only current caller of this from our own code.

But I'd also say my proposed change is not "wrong"; it has pretty defensible API semantics in that just changes the function to work the same as `lcfs_node_set_xattr` does basically (which also currently does not signal to the caller in any way whether there was an extant value).

Anyways though since I don't have a strong feeling on this I'll change it to match your suggestion, thanks!

Reply:
> I think it's the original code that's wrong.

To be clear: I agree :)

Reply:
OK, I changed it.

Though one thing this change makes really obvious is that we desperately need API documentation using a standard tool that does things like describe semantics and especially return error codes. I'm out of the loop on state of the art in tooling there...systemd seems to do man-page-per-function which seems quite heavyweight but maybe is viable for us. Is doxygen still the top thing? (Seems so?)

Reply:
What ever happened to good old gtk-doc? :)